Sources/PenguinStructures/ArrayStorage.swift

```diff
@@ -0,0 +1,184 @@
+  /// insufficient capacity remaining
+  public final func appendValue(at p: UnsafeRawPointer) -> Int? {
+    implementation.appendValue_(at: p)
+  }
+
+  /// Invokes `body` with the memory occupied by initialized elements.
+  public final func withUnsafeMutableRawBufferPointer<R>(
+    _ body: (inout UnsafeMutableRawBufferPointer)->R
+  ) -> R {
+    implementation.withUnsafeMutableRawBufferPointer_(body)
+  }
+
+  /// Deinitializes all stored data.
+  deinit { implementation.deinitialize() }
+}
+
+/// Contiguous storage of homogeneous elements of statically unknown type.
+///
+/// Conformances to this protocol provide the implementations for
+/// `AnyArrayStorage` APIs.
+public protocol AnyArrayStorageImplementation: AnyArrayStorage {
+  /// Appends the instance of the concrete element type whose address is `p`,
+  /// returning the index of the appended element, or `nil` if there was
+  /// insufficient capacity remaining
+  func appendValue_(at p: UnsafeRawPointer) -> Int?
+  
+  /// Invokes `body` with the memory occupied by initialized elements.
+  func withUnsafeMutableRawBufferPointer_<R>(
+    _ body: (inout UnsafeMutableRawBufferPointer)->R
+  ) -> R
+
+  /// Deinitialize stored data
+  func deinitialize()
+}
+
+/// Contiguous storage of homogeneous elements of statically known type.
+///
+/// This protocol's extensions provide APIs that depend on the element type, and
+/// the implementations for `AnyArrayStorage` APIs.
+public protocol ArrayStorageImplementation: AnyArrayStorageImplementation {
+  associatedtype Element
+}
+
+/// APIs that depend on the `Element` type.
+extension ArrayStorageImplementation {
+  /// A type whose instances can be used to access the memory of `self` with a
+  /// degree of type-safety.
+  private typealias Accessor = ManagedBufferPointer<ArrayHeader, Element>
+
+  /// A handle to the memory of `self` providing a degree of type-safety.
+  private var access: Accessor { .init(unsafeBufferObject: self) }
+
+  /// The number of elements stored in `self`.
+  public var count: Int {
+    _read { yield access.withUnsafeMutablePointerToHeader { $0.pointee.count } }
+    _modify {
+      defer { _fixLifetime(self) }
+      yield &access.withUnsafeMutablePointerToHeader { $0 }.pointee.count
+    }
+  }
+
+  /// The maximum number of elements that can be stored in `self`.
+  public var capacity: Int {
+    _read {
+      yield access.withUnsafeMutablePointerToHeader {
+        $0.pointee.capacity }
+    }
+    _modify {
+      defer { _fixLifetime(self) }
+      yield &access.withUnsafeMutablePointerToHeader { $0 }.pointee.capacity
+    }
+  }
+  
+  /// Appends `x` if possible, returning the index of the appended element or
+  /// `nil` if there was insufficient capacity remaining.
+  public func append(_ x: Element) -> Int? {
+    let r = count
+    if r == capacity { return nil }
+    access.withUnsafeMutablePointers { h, e in
+      (e + r).initialize(to: x)
+      h[0].count = r + 1
```

Comment:
Is this the same as:

```suggestion
      h.pointee.count = r + 1
```

Or am I missing something? If so, I think the suggestion might be easier to read.

Reply:
Yes, they're the same.  I don't have a strong opinion about which is easier to read; they each have strengths.  If you have any conviction about this, I'll take your suggestion.  But I wish I could write `h*.count`.  And maybe we should have that operator in penguin internals.

Reply:
+1 for having an operator. I can put together a PR shortly, and I'll update this code as well.

Reply:
We could follow Pascal precedent and have `h^.count`.  I think I prefer `*` tho.